### PR TITLE
prevent schema import failure errors during all imodel-transform usage

### PR DESCRIPTION
**Pull request by MichaelBelousov**
- filter out system schemas from import of imodel transform

### CODE REVIEW

**Comment by MichaelBelousov** [comment]
@scsewall what do you think about this approach? These schemas appear to be default schemas in new iModels (although not system schemas it seems so I'll remove the comment's reference to "system". Do you know of a better name for these schemas and maybe a place to put this filter function, or we could keep it local here.

**Review by scsewall — COMMENTED** [review]
I would prefer to check a schema attribute instead of maintaining a hard-coded list, but I suppose this is an improvement of the current behavior.

Do you get a "duplicate error" if you re-import a regular schema file (like BisCore) when the iModel is already at the latest version of BisCore?

**Comment by MichaelBelousov** [comment]
> Do you get a "duplicate error" if you re-import a regular schema file (like BisCore) when the iModel is already at the latest version of BisCore?

you should but I'll double check

**Comment by scsewall** [comment]
> > Do you get a "duplicate error" if you re-import a regular schema file (like BisCore) when the iModel is already at the latest version of BisCore?
> 
> you should but I'll double check

From an IModelTransformer perspective, that should not cause a "duplicate error" log either...

**Comment by scsewall** [comment]
We should also consider the case where there really was an update to `ECDbMap` or `ECDbSchemaPolicies`.

**Comment by MichaelBelousov** [comment]
yeah I suppose we could check for the schema with a locate before importing it, but I'll look more into it later today

**Comment by MichaelBelousov** [comment]
@scsewall I updated it by exposing `containsSchema`, but the code for checking versions is a bit verbose. It might be nicer if I instead exposed a `locateSchema` function, but I couldn't find a precedent for jsonifying a schema key. Do you know of any, or should we avoid establishing a precedent for that and keep to schema names as I did.

Also I attached the corresponding backend PR to the initial comment.

**Comment by scsewall** [comment]
> @scsewall I updated it by exposing `containsSchema`, but the code for checking versions is a bit verbose. It might be nicer if I instead exposed a `locateSchema` function, but I couldn't find a precedent for jsonifying a schema key. Do you know of any, or should we avoid establishing a precedent for that and keep to schema names as I did.
> 
> Also I attached the corresponding backend PR to the initial comment.

Are you aware of `IModelDb.querySchemaVersion` and the `semver` package?

**Comment by MichaelBelousov** [comment]
> Are you aware of `IModelDb.querySchemaVersion` and the `semver` package?
nope 😄 going to go use that

**Comment by MichaelBelousov** [comment]
looks like '1.00.00' is an invalid syntax according to semver, and that is found in the version string, is there something else I'm missing?

**Comment by scsewall** [comment]
> looks like '1.00.00' is an invalid syntax according to semver, and 0-padding is found in the version string in the file name, so I think I still custom parsing for that part, but maybe I can use comparisons from semver

You are correct. Looks like there is an open request against `semver` on this exact issue. `IModelDb.querySchemaVersion` gets around this by querying the version out as integers and then building a string. We should probably add a utility somewhere to convert a schema version string (that may have leading zeros) to something semver compatible.

**Comment by MichaelBelousov** [comment]
updated with a new check, any idea where I could put the padded version string to semver converter?

**Comment by MichaelBelousov** [comment]
maybe I can just add `paddedVersionToSemver` as a static method to the `Schema` class, maybe there's a better less verbose name though...

**Comment by scsewall** [comment]
> I added `paddedVersionToSemver` as a static method to the `Schema` class, maybe there's a better less verbose name though...

`toSemverString`?

**Comment by MichaelBelousov** [comment]
> `toSemverString`?

my only concern is that doesn't indicate that it's intended to be used on padded version strings like `1.00.00` in the name, if it weren't on the schema class but in some `SchemaVersions` namespace then I think `toSemverString` would be descriptive enough.
If you think it's descriptive enough anyway I can switch it. I'll add some tests.

**Comment by scsewall** [comment]
> > `toSemverString`?
> 
> my only concern is that doesn't indicate that it's intended to be used on padded version strings like `1.00.00` in the name, if it weren't on the schema class but in some `SchemaVersions` namespace then I think `toSemverString` would be descriptive enough.
> Maybe I'm overthinking it, If you think it's descriptive enough anyway I can switch it. I'll add some tests.

My reasoning for the shorter name - the method should be safe to call even if the input string is not padded. You can document the "remove leading zero / padding" intent in the method description instead of bloating up the method name.

**Comment by MichaelBelousov** [comment]
sounds good, I'll switch the name

**Comment by MichaelBelousov** [comment]
looks like this filtering occasionally resulted in *no* schemas getting imported which is an error and one of our tests covered it. I'll just branch to not import if there are no schemas

**Comment by MichaelBelousov** [comment]
I now partially parse the xml files to get the version and schemaName from there.

File: core/backend/src/IModelTransformer.ts [hunk 1]
```diff
@@ -755,7 +757,32 @@ export class IModelTransformer extends IModelExportHandler {
     try {
       this.sourceDb.nativeDb.exportSchemas(schemasDir);
       const schemaFiles: string[] = IModelJsFs.readdirSync(schemasDir);
-      await this.targetDb.importSchemas(requestContext, schemaFiles.map((fileName) => path.join(schemasDir, fileName)));
+      // some schemas are guaranteed to exist and importing them will be a duplicate schema error, so we filter them out
+      const importSchemasFullPaths = schemaFiles.map((schema) => path.join(schemasDir, schema));
+      const filteredSchemaPaths = importSchemasFullPaths.filter((schemaPath) => {
+        let schemaSource = "";
```
Comment (scsewall) [inline core/backend/src/IModelTransformer.ts:763]: Minor nit, but how about `let schemaSource: string;` instead so you don't assign a value that you are going to immediately overwrite.

File: core/backend/src/IModelTransformer.ts [hunk 2]
```diff
@@ -755,7 +757,32 @@ export class IModelTransformer extends IModelExportHandler {
...
+        const schemaNameMatch = /<ECSchema .*?schemaName="([^"]+)"/.exec(schemaSource);
+        if (schemaVersionMatch == null || schemaNameMatch == null) {
+          Logger.logError(loggerCategory, `failed to parse schema name or version, first 200 chars: '${schemaSource.slice(0, 200)}'`);
+          return true;
+        }
+        const [_fullVersionMatch, versionString] = schemaVersionMatch;
+        const [_fullNameMatch, schemaName] = schemaNameMatch;
+        const versionInTarget = this.targetDb.querySchemaVersion(schemaName);
+        const versionToImport = Schema.toSemverString(versionString);
+        if (versionInTarget && Semver.eq(versionToImport, versionInTarget))
```
Comment (scsewall) [inline core/backend/src/IModelTransformer.ts:780]: Don't we want less than or equal?

**Review by scsewall — CHANGES_REQUESTED** [review]

File: core/backend/src/IModelTransformer.ts (hunk 2 above)
Comment (MichaelBelousov) [reply core/backend/src/IModelTransformer.ts:780]: the error occurs only when the exact schema version to be imported is already imported, so the goal is to not import it. We could do it less-than but I don't think there is much of a difference...

File: core/backend/src/IModelTransformer.ts (hunk 1 above)
Comment (MichaelBelousov) [reply core/backend/src/IModelTransformer.ts:763]: that's better. I'll change it.

File: core/backend/src/IModelTransformer.ts (hunk 2 above)
Comment (scsewall) [reply core/backend/src/IModelTransformer.ts:780]: Not an issue of functionality, but from a performance perspective it seems wasteful to make the native code side open the file again and probably parse the xml.

**Review by scsewall — APPROVED** [review]

**Pull request merged**